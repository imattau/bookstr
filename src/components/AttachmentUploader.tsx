import React, { useState } from 'react';
import { useNostr } from '../nostr';

async function uploadAttachment(file: File): Promise<string> {
  const fd = new FormData();
  fd.append('file', file);
  const res = await fetch('/api/upload', {
    method: 'POST',
    body: fd,
  });
  if (!res.ok) throw new Error('upload failed');
  const data = await res.json();
  return data.url as string;
}

interface AttachmentUploaderProps {
  bookId?: string;
  onUploaded?: (url: string) => void;
}

export const AttachmentUploader: React.FC<AttachmentUploaderProps> = ({
  bookId,
  onUploaded,
}) => {
  const ctx = useNostr();
  const [loading, setLoading] = useState(false);

  const handleFile = async (
    e: React.ChangeEvent<HTMLInputElement>,
  ) => {
    const file = e.target.files?.[0];
    if (!file) return;
    setLoading(true);
    try {
      const url = await uploadAttachment(file);
      await ctx.publish({
        kind: 1064,
        content: '',
        tags: [
          ['mime', file.type],
          ['url', url],
          ...(bookId ? [['e', bookId]] : []),
        ],
      });
      onUploaded?.(url);
    } catch {
      // ignore errors
    } finally {
      setLoading(false);
      e.target.value = '';
    }
  };

  return (
    <div className="space-y-2">
      <input type="file" onChange={handleFile} disabled={loading} />
      {loading && <p className="text-sm">Uploading...</p>}
    </div>
  );
};
