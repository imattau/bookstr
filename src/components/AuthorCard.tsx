import React from 'react';
import { FollowButton } from './FollowButton';

export interface AuthorCardProps {
  pubkey: string;
  name?: string;
}

/**
 * Simple author card showing a name/pubkey with follow button.
 */
export const AuthorCard: React.FC<AuthorCardProps> = ({ pubkey, name }) => {
  return (
    <div className="flex items-center gap-[var(--space-2)] rounded-[var(--radius-card)] border p-[var(--space-2)]">
      <div className="flex-1">
        <p className="font-medium">{name || pubkey}</p>
      </div>
      <FollowButton pubkey={pubkey} />
    </div>
  );
};
