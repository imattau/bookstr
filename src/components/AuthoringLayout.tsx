import React from 'react';

export interface AuthoringLayoutProps {
  sidebar: React.ReactNode;
  content: React.ReactNode;
}

/**
 * Layout used for authoring screens with optional sidebar.
 */
export const AuthoringLayout: React.FC<AuthoringLayoutProps> = ({ sidebar, content }) => {
  return (
    <div className="mx-auto max-w-5xl px-4 py-6">
      <div className="grid gap-6 lg:grid-cols-[minmax(12rem,1fr)_3fr]">
        <aside className="hidden lg:block">{sidebar}</aside>
        <main>{content}</main>
      </div>
    </div>
  );
};
