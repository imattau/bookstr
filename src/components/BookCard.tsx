import React, { useState } from 'react';
import { useNostr, zap } from '../nostr';
import type { Event as NostrEvent } from 'nostr-tools';
import { logEvent } from '../analytics';

interface BookCardProps {
  event: NostrEvent;
}

export const BookCard: React.FC<BookCardProps> = ({ event }) => {
  const title = event.tags.find((t) => t[0] === 'title')?.[1] ?? 'Untitled';
  const summary = event.tags.find((t) => t[0] === 'summary')?.[1] ?? '';
  const cover = event.tags.find((t) => t[0] === 'image')?.[1];
  const ctx = useNostr();
  const [status, setStatus] = useState<'idle' | 'zapping' | 'done'>('idle');
  const { toggleBookmark, bookmarks } = ctx;

  const handleZap = async () => {
    setStatus('zapping');
    try {
      await zap(ctx, event);
      logEvent('click_vote', { id: event.id });
      setStatus('done');
    } catch {
      setStatus('idle');
    }
  };

  const handleFav = async () => {
    try {
      await toggleBookmark(event.id);
      logEvent('click_fav', { id: event.id });
    } catch {
      /* ignore */
    }
  };

  return (
    <div className="rounded border p-2">
      {cover && (
        <img src={cover} alt="" className="mb-2 h-32 w-24 object-cover" />
      )}
      <h3 className="font-semibold">{title}</h3>
      {summary && <p className="text-sm text-gray-500">{summary}</p>}
      <div className="pt-2 flex gap-2">
        <button
          onClick={handleZap}
          className="rounded bg-yellow-500 px-2 py-1 text-white"
        >
          {status === 'zapping'
            ? 'Zapping...'
            : status === 'done'
              ? 'Zapped!'
              : 'Zap'}
        </button>
        <button
          onClick={handleFav}
          aria-label="Favorite"
          className="rounded border px-2 py-1"
        >
          â˜…
        </button>
      </div>
    </div>
  );
};
