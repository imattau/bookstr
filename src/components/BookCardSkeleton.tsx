import React from 'react';
import { Skeleton, TextSkeleton } from './Skeleton';

export const BookCardSkeleton: React.FC = () => (
  <div className="rounded-[var(--radius-card)] border p-[var(--space-2)] space-y-[var(--space-2)]">
    <Skeleton className="h-32 w-24" />
    <TextSkeleton lines={1} className="w-3/4" />
    <TextSkeleton lines={1} />
    <div className="pt-2 flex gap-2">
      <Skeleton className="h-6 w-16 rounded" />
      <Skeleton className="h-6 w-8 rounded" />
    </div>
  </div>
);
