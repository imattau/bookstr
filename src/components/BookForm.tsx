import React from 'react';
import { useNostr } from '../nostr';

export const BookForm: React.FC = () => {
  const { publish } = useNostr();
  const [title, setTitle] = React.useState('');
  const [content, setContent] = React.useState('');
  const [summary, setSummary] = React.useState('');
  const [tags, setTags] = React.useState('');
  const [cover, setCover] = React.useState('');

  const handlePublish = async () => {
    await publish({
      kind: 30023,
      content,
      tags: [
        ['title', title],
        ['summary', summary],
        ...tags
          .split(',')
          .map((t) => t.trim())
          .filter(Boolean)
          .map((t) => ['t', t]),
        ...(cover ? [['image', cover]] : []),
      ],
    });
    setTitle('');
    setSummary('');
    setContent('');
    setTags('');
    setCover('');
  };

  return (
    <div className="space-y-2">
      <input
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        placeholder="Title"
        className="w-full rounded border p-2"
      />
      <textarea
        value={summary}
        onChange={(e) => setSummary(e.target.value)}
        placeholder="Summary"
        className="w-full rounded border p-2"
      />
      <textarea
        value={content}
        onChange={(e) => setContent(e.target.value)}
        placeholder="Content"
        className="w-full rounded border p-2"
      />
      <input
        value={tags}
        onChange={(e) => setTags(e.target.value)}
        placeholder="Tags comma separated"
        className="w-full rounded border p-2"
      />
      <input
        value={cover}
        onChange={(e) => setCover(e.target.value)}
        placeholder="Cover URL"
        className="w-full rounded border p-2"
      />
      <button
        onClick={handlePublish}
        className="rounded bg-primary-600 px-4 py-2 text-white"
      >
        Publish
      </button>
    </div>
  );
};
