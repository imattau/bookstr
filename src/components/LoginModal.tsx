import React from 'react';
import { useNostr } from '../nostr';
import { connectNostrWallet, nostrLogin } from '../nostr/auth';
import { importKey, validatePrivKey } from '../lib/keys';
import { isValidUrl } from '../validators';

export interface LoginModalProps {
  onClose: () => void;
}

export const LoginModal: React.FC<LoginModalProps> = ({ onClose }) => {
  const ctx = useNostr();
  const { login, logout, pubkey } = ctx;
  const [tab, setTab] = React.useState<'priv' | 'nip07' | 'remote'>('priv');

  const [privInput, setPrivInput] = React.useState('');
  const [privError, setPrivError] = React.useState<string | null>(null);

  const [remoteUrl, setRemoteUrl] = React.useState('');
  const [remoteToken, setRemoteToken] = React.useState('');
  const [remoteError, setRemoteError] = React.useState<string | null>(null);
  const [loading, setLoading] = React.useState(false);

  const hasWallet = React.useMemo(() => Boolean((window as any).nostr), []);

  const handlePrivLogin = () => {
    const key = importKey(privInput);
    if (!key || !validatePrivKey(key)) {
      setPrivError('Invalid private key');
      return;
    }
    try {
      login(key);
      setPrivError(null);
      onClose();
    } catch {
      setPrivError('Invalid private key');
    }
  };

  const handleWalletLogin = async () => {
    setLoading(true);
    try {
      const pk = await connectNostrWallet();
      if (!pk) throw new Error('Wallet not found');
      await nostrLogin(ctx, pk);
      onClose();
    } catch (err: any) {
      setPrivError(err.message);
    } finally {
      setLoading(false);
    }
  };

  const handleRemoteLogin = async () => {
    setLoading(true);
    try {
      const urlOk = isValidUrl(remoteUrl);
      if (!urlOk || !remoteToken) {
        setRemoteError('Enter a valid URL and token');
        setLoading(false);
        return;
      }
      const res = await fetch(`${remoteUrl.replace(/\/$/, '')}/pubkey`, {
        headers: { Authorization: `Bearer ${remoteToken}` },
      });
      const data = await res.json();
      if (!data.pubkey) throw new Error('Invalid response');
      ctx.loginNip07(data.pubkey);
      localStorage.setItem('remoteSignerUrl', remoteUrl);
      localStorage.setItem('remoteSignerToken', remoteToken);
      setRemoteError(null);
      onClose();
    } catch (err: any) {
      setRemoteError(err.message);
    } finally {
      setLoading(false);
    }
  };

  const isPrivValid = Boolean(importKey(privInput));
  const isRemoteValid = isValidUrl(remoteUrl) && Boolean(remoteToken);

  if (pubkey) {
    return (
        <div className="fixed inset-0 z-50 flex items-center justify-center bg-black/50 p-[var(--space-2)]">
          <div className="w-full max-w-sm space-y-4 rounded bg-[color:var(--clr-surface)] p-[var(--space-4)]">
          <p className="text-sm break-all">Logged in as {pubkey}</p>
          <div className="flex justify-end gap-2">
              <button onClick={logout} className="rounded border px-[var(--space-3)] py-[var(--space-1)]">
              Logout
            </button>
              <button onClick={onClose} className="rounded bg-[color:var(--clr-primary-600)] px-[var(--space-3)] py-[var(--space-1)] text-white">
              Close
            </button>
          </div>
        </div>
      </div>
    );
  }

  return (
      <div className="fixed inset-0 z-50 flex items-center justify-center bg-black/50 p-[var(--space-2)]">
        <div className="w-full max-w-sm space-y-4 rounded bg-[color:var(--clr-surface)] p-[var(--space-4)]">
        <div className="flex gap-2">
          <button
            onClick={() => setTab('priv')}
            className={`flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] ${tab === 'priv' ? 'bg-border' : ''}`}
          >
            Private Key
          </button>
          <button
            onClick={() => setTab('nip07')}
            className={`flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] ${tab === 'nip07' ? 'bg-border' : ''}`}
          >
            NIP-07
          </button>
          <button
            onClick={() => setTab('remote')}
            className={`flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] ${tab === 'remote' ? 'bg-border' : ''}`}
          >
            Remote
          </button>
        </div>
        {tab === 'priv' && (
          <div className="space-y-2">
            <input
              value={privInput}
              onChange={(e) => setPrivInput(e.target.value)}
              placeholder="nsec or hex"
              className="w-full rounded border p-[var(--space-2)]"
            />
            {privError && <p className="text-red-600 text-sm">{privError}</p>}
            <button
              onClick={handlePrivLogin}
              disabled={!isPrivValid || loading}
              className="w-full rounded bg-[color:var(--clr-primary-600)] px-[var(--space-3)] py-[var(--space-1)] text-white disabled:opacity-50"
            >
              Login
            </button>
          </div>
        )}
        {tab === 'nip07' && (
          <div className="space-y-2">
            {!hasWallet && <p className="text-sm">Nostr wallet not detected.</p>}
            {privError && <p className="text-red-600 text-sm">{privError}</p>}
            <button
              onClick={handleWalletLogin}
              disabled={!hasWallet || loading}
              className="w-full rounded bg-[color:var(--clr-primary-600)] px-[var(--space-3)] py-[var(--space-1)] text-white disabled:opacity-50"
            >
              Connect Wallet
            </button>
          </div>
        )}
        {tab === 'remote' && (
          <div className="space-y-2">
            <input
              value={remoteUrl}
              onChange={(e) => setRemoteUrl(e.target.value)}
              placeholder="Signer URL"
              className="w-full rounded border p-[var(--space-2)]"
            />
            <input
              value={remoteToken}
              onChange={(e) => setRemoteToken(e.target.value)}
              placeholder="Auth token"
              className="w-full rounded border p-[var(--space-2)]"
            />
            {remoteError && <p className="text-red-600 text-sm">{remoteError}</p>}
            <button
              onClick={handleRemoteLogin}
              disabled={!isRemoteValid || loading}
              className="w-full rounded bg-[color:var(--clr-primary-600)] px-[var(--space-3)] py-[var(--space-1)] text-white disabled:opacity-50"
            >
              Connect
            </button>
          </div>
        )}
        <div className="flex justify-end">
          <button onClick={onClose} className="rounded px-[var(--space-3)] py-[var(--space-1)] border">
            Close
          </button>
        </div>
      </div>
    </div>
  );
};
