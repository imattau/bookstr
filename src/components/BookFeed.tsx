import React, { useEffect, useState } from 'react';
import { useNostr } from '../nostr';
import { BookCard } from './BookCard';
import { BookCardSkeleton } from './BookCardSkeleton';
import type { Event as NostrEvent } from 'nostr-tools';
import { useEventStore } from '../store/events';

/**
 * Feed showing recent book publications from followed relays.
 */
export const BookFeed: React.FC = () => {
  const { subscribe } = useNostr();
  const addEvent = useEventStore((s) => s.addEvent);
  const [events, setEvents] = useState<
    (NostrEvent & { repostedBy?: string })[]
  >([]);

  useEffect(() => {
    const offMain = subscribe([{ kinds: [30023], limit: 20 }], (evt) => {
      addEvent(evt);
      setEvents((e) => {
        const d = evt.tags.find((t) => t[0] === 'd')?.[1];
        if (d) {
          const idx = e.findIndex(
            (x) =>
              x.kind === 30023 &&
              x.pubkey === evt.pubkey &&
              x.tags.find((t) => t[0] === 'd')?.[1] === d,
          );
          if (idx !== -1) {
            if (e[idx].created_at >= evt.created_at) return e;
            const copy = [...e];
            copy[idx] = evt;
            return copy;
          }
        }
        if (e.find((x) => x.id === evt.id)) return e;
        return [...e, evt];
      });
    });
    const offRepost = subscribe([{ kinds: [6], limit: 20 }], (evt) => {
      const target = evt.tags.find((t) => t[0] === 'e')?.[1];
      if (!target) return;
      const offTarget = subscribe([{ ids: [target] }], (orig) => {
        addEvent(orig);
        setEvents((e) => {
          const d = orig.tags.find((t) => t[0] === 'd')?.[1];
          if (d) {
            const idx = e.findIndex(
              (x) =>
                x.kind === 30023 &&
                x.pubkey === orig.pubkey &&
                x.tags.find((t) => t[0] === 'd')?.[1] === d,
            );
            if (idx !== -1) {
              if (e[idx].created_at >= orig.created_at) {
                const copy = [...e];
                copy[idx] = { ...copy[idx], repostedBy: evt.pubkey };
                return copy;
              }
              const copy = [...e];
              copy[idx] = { ...orig, repostedBy: evt.pubkey };
              return copy;
            }
          }
          const idx = e.findIndex((x) => x.id === orig.id);
          if (idx !== -1) {
            const copy = [...e];
            copy[idx] = { ...copy[idx], repostedBy: evt.pubkey };
            return copy;
          }
          return [...e, { ...orig, repostedBy: evt.pubkey }];
        });
        offTarget();
      });
    });
    return () => {
      offMain();
      offRepost();
    };
  }, [subscribe, addEvent]);

  return (
    <ul role="list" className="space-y-4">
      {events.length === 0
        ? Array.from({ length: 3 }).map((_, i) => (
            <li key={i} role="listitem">
              <BookCardSkeleton />
            </li>
          ))
        : events.map((e) => (
            <li key={e.id} role="listitem">
              <BookCard
                event={e}
                onDelete={(id) =>
                  setEvents((evts) => evts.filter((x) => x.id !== id))
                }
              />
            </li>
          ))}
    </ul>
  );
};
