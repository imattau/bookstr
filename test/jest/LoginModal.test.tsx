import React from 'react';
import { render } from '@testing-library/react';
import { LoginModal } from '../../src/components/LoginModal';

jest.mock('../../src/components/ToastProvider', () => ({
  useToast: () => jest.fn(),
}));

jest.mock('../../src/nostr', () => ({
  useNostr: () => ({
    login: jest.fn(),
    logout: jest.fn(),
    loginNip07: jest.fn(),
    pubkey: null,
  }),
}));

jest.mock('../../src/nostr/auth', () => ({
  connectNostrWallet: jest.fn(),
  nostrLogin: jest.fn(),
}));

describe('LoginModal', () => {
  it('matches snapshot', () => {
    const { container } = render(<LoginModal onClose={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
