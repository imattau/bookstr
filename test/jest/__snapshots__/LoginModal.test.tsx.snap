// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoginModal matches snapshot 1`] = `
<div
  class="fixed inset-0 z-50 flex items-center justify-center bg-black/50 p-[var(--space-2)]"
>
  <div
    class="w-full max-w-sm space-y-4 rounded bg-[color:var(--clr-surface)] p-[var(--space-4)]"
  >
    <div
      class="flex gap-2"
    >
      <button
        class="flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] bg-border"
      >
        Private Key
      </button>
      <button
        class="flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] "
      >
        NIP-07
      </button>
      <button
        class="flex-1 rounded border px-[var(--space-2)] py-[var(--space-1)] "
      >
        Remote
      </button>
    </div>
    <div
      class="space-y-2"
    >
      <div
        class="flex gap-2"
      >
        <input
          class="flex-1 rounded border p-[var(--space-2)]"
          placeholder="nsec or hex"
          type="password"
          value=""
        />
        <button
          aria-label="Show private key"
          class="rounded border px-[var(--space-2)]"
          type="button"
        >
          Show
        </button>
      </div>
      <button
        class="w-full rounded bg-[color:var(--clr-primary-600)] px-[var(--space-3)] py-[var(--space-1)] text-white disabled:opacity-50"
        disabled=""
      >
        Login
      </button>
    </div>
    <div
      class="flex justify-end"
    >
      <button
        class="rounded px-[var(--space-3)] py-[var(--space-1)] border"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
